### DISCUSSION TITLE
Dictionaries accept int as key, unifying arrays, dictionaries and function calls

### DISCUSSION THREAD

**Discussion by uben0** [typst/typst#5948] [Ideas]
## What if dictionary keys can be integers?

```typst
#let foo = (0: [Hello World], lang: "en")
#let bar = ([Hello World], lang: "en")
```

Then, expansion could affect positional arguments:

```typst
#text(.. foo)
#text(0: [Hello World], lang: "en")
#text([Hello World], lang: "en")
```

### Simpler construct

With this, the call to a function is not a specific language construct anymore, but just a `function + dictionay`.

This "should not be" a breaking change.

## Use Cases

The most naïve use case, is to store redundant parameters of repeated function call.

```typst
#let p = (left, ltr, stroke: 1pt + blue)

#bubble("Alice", .. p)
#bubble("Bob", .. p)
#bubble("Charlie", .. p)
```

### `set` instruction

Which could also be a nice definition of the `set` instruction, that, by the way, could be extended to all functions:

```typst
#set bubble(1: left, 2: ltr, stroke: 1pt + blue)

#let bubble-alias(name) = bubble(name, left, ltr, stroke: 1pt + blue)
#let bubble = bubble-alias
```
### Typed vs non-typed

Match the implicit *constructor pattern* used, for instance, in CeTZ, where some (typed) data is given as a tree of arrays and dictionaries, that are used to construct the type, like a position:

```typst
#let pos1 = (rel: (1, 0))
#let pos2 = ("a",)
#let pos3 = (rel: "a", to: (0, -2))
```

We could see functions as constructors accepting a single input: **a str/int keyed dictionary**

Here, `pos1`, `pos2` and `pos3` are just potential inputs, they don't have a type (except being arrays and dicts). They would become typed once fed to an hypothetical `position` function/type/constructor.

### `show` instruction

It would make the `show` instruction more flexible, as it would give `it` as non-typed data (the bare input dictionary).

```typst
#show heading: it => repr(it)

== Hello
```
```
(0: [Hello], level: 2)
```

**Comment by uben0** [comment]
I am proposing this, because in my understanding (which could be limited), it removes complexity from the language, it doesn't add anything (the best part is no part).

Am I correct when saying it removes complexity? Do you see my point? Am I wrong?

**Comment by laurmaedje** [comment]
Let me try to respond to all of your individual points.

### Simpler construct
> With this, the call to a function is not a specific language construct anymore, but just a function + dictionay.

This here I don't quite understand. Right now, it is similarly just `function + arguments` ([arguments type](https://typst.app/docs/reference/foundations/arguments/)).

### Repeated parameters
> The most naïve use case, is to store redundant parameters of repeated function call.

I agree that the merged type is very nice for the argument use. It would unify the types `array`, `dictionary`, and `arguments` here. It is similar to Lua tables. Note that your example is also quite simple to write currently though: 
```typ
#let p = arguments(left, ltr, stroke: 1pt + blue)

#bubble("Alice", .. p)
#bubble("Bob", .. p)
#bubble("Charlie", .. p)
```

I think we need to distinguish two separate proposals:
- Dictionary keys can take more values, e.g. integers, but this does not mean that they take over the role of arrays in the language.
- Dictionaries can have integer keys and replace the role typically taken by arrays.

I think the former would be desirable just as a way to have a sparse int -> value mapping, but the latter would probably not be a great experience. 

Typst actually did originally have a unified table type instead of array and dictionary a long time ago. However, I fear that the price for this unification is complexity in other places. In my experience back then, it wasn't a great experience. The two data types are typically used for different operations and representing an array as a dictionary with integer keys raises all sorts of questions about what happens when a key is missing, etc.

### Set rules
> Which could also be a nice definition of the set instruction, that, by the way, could be extended to all functions:

Why `#set bubble(1: left, 2: ltr, stroke: 1pt + blue)` instead of just `#set bubble(left, ltr, stroke: 1pt + blue)`?

### Typed vs non-typed
> Match the implicit constructor pattern used, for instance, in CeTZ, where some (typed) data is given as a tree of arrays and dictionaries, that are used to construct the type, like a position:

It's true that dictionaries are currently used a lot for configuration and the lack of mixing of positional and named properties might be limiting. I think one other way to improve this is to use custom types for those things once they are available. Overall, the plan is to go more into the direction of making element functions proper types and then show & set operate on types rather than functions.

### Show rules
> It would make the show instruction more flexible, as it would give it as non-typed data (the bare input dictionary).

It's important for `it` to be something more than just a dictionary because you can use `it` itself to invoke the next outer show rule. `it` can also hold other important data like a label.

**Comment by uben0** [comment]
Oh I see!

You just introduced me to the `arguments` type. I didn't know it existed. Which seems to be basically what I proposed here.

## Array, Dictionary and Arguments

Currently, `(a, b, c)` is an array, `(x: a, y: b, z: c)` is a dictionary, this is syntactically explicit. Each of them can be promoted to the `arguments` type, with for instance `arguments(.. array, .. dictionary)`.

Then, I would propose the promotion to `arguments` to be automatic when the array syntax and dictionary syntax are mixed.

```typst
#let p = (a, x: b)
#type(p) // arguments
```

- **Pros:** nice, especially in the use case exposed with *CeTZ* (allows neater API)
- **Cons:** if the user forgets a tag, they ends up with `error: expected dictionary, found arguments`


## Sparse Arguments

Now lets answer your question:

> Why #set bubble(1: left, 2: ltr, stroke: 1pt + blue) instead of just #set bubble(left, ltr, stroke: 1pt + blue)?

Because we might not want to assign the first argument. As `set` works with named parameters, it's possible to pick the ones we want, it could be the same with positions.

At the moment, `arguments` can't be sparse. But let's imagine it can with the `_` placeholder:

```typst
#let p = arguments(a, _, x: b)
#foo(.. p) // error: missing positional argument
#foo(.. p, c) // ok
```

But nope, bad idea. **This would be quite ambiguous.** What happens when several sparse `arguments` are used as sink?

## Partial application

But what if, calling a function with a sparse `arguments` produces a partial application?

```typst
#let foo = foo(a, _, x: b)
#foo(c)
```
Replacing, by the way, the use of `set` instruction. (Or would it? Not sure about how `set` works.)

Then, using sparse `arguments` in a sink would not be ambiguous:

```typst
#let p1 = (_, a)
#let p2 = (_, _, x: b)
#let p = (.. p1, c, ..p2)
#repr(p)
```
```
(_, a, c, _, _, x: b)
```

**Comment by tingerrr** [comment]
> Then, I would propose the promotion to arguments to be automatic when the array syntax and dictionary syntax are mixed.

I generally like the idea, but that would indeed cause confusion in some cases, and the question is if there are enough potential uses for this to justify this.

> Replacing, by the way, the use of set instruction. (Or would it? Not sure about how set works.)

Nope, consider:

```typst
#set text(fill: red)

#include "x.typ" // this will be red
```
vs
```typst
#let text = text(fill: red, _)

#include "x.typ" // this won't, text is not redefined here
```

Partial application currently works pretty well using `.with(...)`, although it's a little more involved if you want to partially apply a positional that isn't the first, I think this is fine as it's a rare thing to partially apply the second argument.

EDIT: On the other hand, I really like the idea of simplifying partial application.

```typst
#let func(a, b) = {
  // complicated logic where the order of a and b matters
}

// now
#array.filter(func.with(1))
#array.filter(x => func(1, x))

#array.filter(x => func(x, 1))

// with better partial application
#array.filter(func(1, _))
#array.filter(func.with(1, _))

#array.filter(func(_, 1))
#array.filter(func.with(_, 1))
```

Probably doesn't justify the added complexity, though.

**Comment by uben0** [comment]
Once again, it was a skill issue from my part, I didn't knew `with` existed. Knowing that, you are right:

> Probably doesn't justify the added complexity, though.

Nonetheless, support for sparse arguments could be added without being a breaking change:

```typst
#let a = arguments(1, _, 3)
#a.pos() // (1, none, 3)
#a.blank() // (false, true, false)
```

Allowing this:

```typst
#array.filter(func.with(_, 1))
```

It would also make, syntactically, a *pattern* a subset of an *arguments*. In other words, in the grammar, the *pattern* symbol could be replaced by the *arguments* symbol. Well, not exactly, as a pattern also include a single identifier or function declaration:

```typst
#let a = 1
#let b(c) = c
```

## Let's explore, just because it's fun and instructive

I have more ideas and reflections to develop about Typst's syntax. Those are bold suggestions I don't expect to be taken seriously, but rather to help understanding the current syntax.

### What about forbidding `(...)` for expression atomization

```typst
#let a = (2 + 3) * 4
#let b = {2 + 3} * 4
```

The two forms are equivalent (well, not exactly, but I'll talk about it later). Removing the use of the former notation could open a new semantic for it:

```typst
#let x = (1,)
#let y = (1)
```

It would become a single item array notation. And could be accepted as `if`, `while` and `for` body:

```typst
#for i in range(0, 3) {(i,)}
#for i in range(0, 3) (i) // (0, 1, 2)
```

```typst
#if true (1) // (1,)
#if false (1) // ()
```

Making it clear, that, when a `( ... )` is used, a collection is created (either an array, dictionary or arguments).

### The optional `;` separator

As I hinted before, `(2 + 3) * 4` and `{2 + 3} * 4` are not strictly equivalent, as in code block, a line break is a separator, while in groups, it isn't.

```typst
#let a = (3 // ok
  + 2) * 4
#let b = {3 // error
  + 2} * 4
```

The two exceptions are:

```typst
#{if a {}
  else {}}
#{a
    .first()}
```

But what if there were more exceptions?

```typst
#{ 1
 + 2
 + 3
  } // ELM's style

#{ 1 +
   2 +
   3 } // compact style

#{
  1
    + 2
    + 3
} // indented style
```

But is there a logic here? Yes, if we only consider a line break to be a separator between two identically indented instructions:

```typst
#{
  a // indent = 2
  b // indent = 2
} // separated

#{a // no indent
  b // indent = 2
} // joined

#{
  a // indent = 2
    b // indent = 4
} // joined
```

But can we go further ?

### No need for `;` nor `\n`

In my understanding, there is only one operator having both a unary (prefixed) and binary form. The `+` and `-` operators. Which is the only reason to have a separator in the first place. If sequence of expression where allowed in the grammar, the following situations would be ambiguous:

```typst
#{2 + 3}
```
Two valid interpretations:
```
(block (add 2 3))
(block 2 (sign '+' 3))
```
But if precedence of the unary form is lower than the binary one, then we don't need a separator anymore. The code block would also be line break insensitive!

Well, there is one more ambiguity with the `set ... : if ...` notation.

### We need `;` and `\n`

Allowing expressions to be sequential removes a nice property from the grammar: The parsing is no longer guaranteed to have linear complexity. (spoiler alert: it can also increase by 10x the size of the parsing graph used by the parser)

Having `\n` replacing `;` only when two instructions have the same indentation seems to cover 99% of the cases, with the 1% having to use the explicit `;`. And it would be a nice contract with the user: "I allow you to omit the `;` if your instructions are aligned"

**Comment by uben0** [comment]
# Interfaces

If merging the `array` and the `dictionary` types in a single `arguments` type could cause annoying ambiguity, the use of interfaces could solve that:

Let's say:
- `arguments` is a type
- `array` is an interface
- `dictionary` is an interface

An `arguments` value would have the `array` interface if it doesn't have named arguments.

An `arguments` value would have the `dictionary` interface if it doesn't have positional arguments.

With a curious situation when the `arguments` value is empty.

We could test if a value have an interface with `in` operator.

```typst
#let foo(..a) = (
  a in dictionary,
  a in array,
)

#foo(1) // (false, true)
#foo(x: 1) // (true, false)
#foo(1, x: 1) // (false, false)
#foo() // (true, true)
```

```typst
#for i in (1, 2) [#i] // 12
#for (k, v) in (x: 1, y: 2) [#k#v] // x1y2
#for n in (1, x: 1) [#n] // error, doesn't have `array` nor `dictionary` interface
#for n in () [#n] // when the two are available, one could be chosen by default
```

The `arguments` type would still have a `pos` and `named` method that strip one kind of item.

**Comment by tingerrr** [comment]
> What about forbidding (...) for expression atomization

I actually suggested that once, but I think it's a little too alien to most users, as well as a huge breaking change.

> If merging the array and the dictionary types in a single arguments type could cause annoying ambiguity, the use of interfaces could solve that:

I'm not sure if typst is the kind of language that would benefit from the added complexity of interfaces, but perhaps it can. Currently, elements all have a common interface such as `.func()`, `.at()` and `.has()`. But with the possibility of the unification of content and values, this may not be needed.

> We could test if a value have an interface with in operator.

There's probably more merit in adding an operator that more clearly illustrates what it does (`a is dictionary`).

**Comment by uben0** [comment]
> little too alien to most users, as well as a huge breaking change.

Yes, I know, definitely alien, definitely breaking. That's why I knew it would not be taken seriously. But damn, that would satisfy soo much the people obsessed with symmetries (i.e me, and all Rust fanatics XD).

But I would argue, Typst is still unstable (hasn't reach 1.0) and have introduced breaking changes on every releases until now. And Typst is already alien. I tried to introduced it to friends (I enable the "missionary proselytizer" mode) and I noticed they were confused about the notation anyway. I guess, it might actually make the syntax less alien if `(...)` always encode a collection, and never "atomization". It would also be clear what the `{...}` means, it is code block, and the only way to "atomize".

> adding an operator that more clearly illustrates

Maybe, instead of adding a specific operator (breaking change). A method on interfaces would do the job, for instance `dicitonary.is(a)`. In the same way partial application is already available through `.with()` method, and don't require a specific syntax.


PS: if a breaking change is too drastic, like the change of semantic of `(...)` notation. It can always be introduce in two steps. First, deprecation. The compiler may emit a warning, and LSP offers a refactor towards new notation. Secondly, after 1~2 years of deprecation, migrate towards new semantic.

**Comment by mattfbacon** [comment]
FYI, this is not ELM style:

```typst
#{ 1 +
   2 +
   3 }
```

This is:

```typst
{ 1
+ 2
+ 3
}
```

**Comment by uben0** [comment]
If types are a thing in future Typst versions, having interfaces could simplify things. For instance, `content` should not be a type but an interface. When giving an integer as content, it would be accepted as content while remain an integer. A newly defined type could implement the content interface to be accepted as content.

**Comment by laurmaedje** [comment]
I have converted this into a discussion because that's what it evolved into. There are various interesting discussions and observations here, but not really a focussed proposal.
